src/device-manager/java/src/nl/Weave/DataManagement/GenericTraitUpdatableDataSinkImpl.java

```diff
@@ -0,0 +1,444 @@
+/*
+
+    Copyright (c) 2019 Google, LLC.
+    All rights reserved.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+ */
+
+
+package nl.Weave.DataManagement;
+
+import android.os.Build;
+import android.util.Log;
+
+import java.math.BigInteger;
+import java.util.EnumSet;
+import java.util.Random;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Iterator;
+
+public class GenericTraitUpdatableDataSinkImpl implements GenericTraitUpdatableDataSink
+{
+    protected GenericTraitUpdatableDataSinkImpl(long traitInstancePtr, WDMClientImpl wdmClient)
+    {
+        mWDMClient = wdmClient;
+        mTraitInstancePtr = traitInstancePtr;
+        mCompHandler = null;
+        init(mTraitInstancePtr);
+    }
+
+    protected void close()
+    {
+        if (mWDMClient != null)
+        {
+            mWDMClient.removeDataSinkRef(mTraitInstancePtr);
+            mWDMClient = null;
+        }
+
+        if (mTraitInstancePtr != 0)
+        {
+            close(mTraitInstancePtr);
+            mTraitInstancePtr = 0;
+        }
+        mCompHandler = null;
+    }
+
+    @Override
+    public void setSigned(String path, int value, boolean isConditional)
+    {
+        boolean isSigned = true;
```

Comment:
Nothing wrong with this so feel free to ignore, but another common pattern is:

```
setInt(mTraitInstancePtr, path, value, isConditional, /* isSigned= */ true);
```

Reply:
@jaylogue what do you think of? using this way, we can remove several similar functions for integers, even for integer array.

Reply:
I might have been misunderstood here. I just meant instead of having a local boolean, inlining the value with a comment. i.e. instead of:

```
boolean isSigned = true;
setInt(mTraitInstancePtr, path, value, isConditional, isSigned);
```

You can just write:

```
setInt(mTraitInstancePtr, path, value, isConditional, /* isSigned= */ true);
```

Reply:
Done, thanks